### ISSUE TITLE
API cleanup suggestions

### ISSUE THREAD

**Issue by james-lawrence** [anacrolix/torrent#243]
as discussed in #218, the API is a little hard to follow and use for certain use cases. The goal here is to discuss and suggest usability improvements to the API. I believe these changes could also be used to help resolve #238, #168 in a clean fashion.

- remove torrent creation methods from the client. there are just want too many of them. poorly documented, and its unclear which one should be used and when.

```golang
func (cl *Client) AddMagnet(uri string) (T *Torrent, err error) {}
func (cl *Client) AddTorrentFromFile(filename string) (T *Torrent, err error) {}
func (cl *Client) AddTorrent(mi *metainfo.MetaInfo) (T *Torrent, err error) {}
func (cl *Client) AddTorrentInfoHash(infoHash metainfo.Hash) (t *Torrent, new bool) {}
func (cl *Client) AddTorrentInfoHashWithStorage(infoHash metainfo.Hash, specStorage storage.ClientImpl) (t *Torrent, new bool) {}
func (cl *Client) AddTorrentSpec(spec *TorrentSpec) (t *Torrent, new bool, err error) {}
```

- Create new standalone methods for the creation of torrent specifications. these methods have no concept of the client and just know how to create torrent specs.

```golang
type TorrentSpecOption func(*TorrentSpec)
// TorrentSpecData provide the source data to the torrent, nil can represent no local data.
type TorrentSpecData(d TorrentImpl) TorrentSpecOption {
  return func(t *TorrentSpec) {
    t.TorrentImpl = d
  }
}

// New create a torrent from the metainfo.Info and any additional options.
func New(mi metainfo.Info, options ...TorrentSpecOption) (*TorrentSpec, error) {}

// NewFromData creates a torrent from the raw data the torrent will represent.
func NewFromData(r io.Reader, options ...TorrentSpecOption) (*TorrentSpec, error) {}

// NewFromFile convience method to create a torrent from a file. see NewFromData for
// more information.
func NewFromFile(path string, options ...TorrentSpecOption) (*TorrentSpec, err error) {}

// NewFromMetadata creates a torrent from metadata stored in a file based format.
func NewFromMetadata(r io.ReadCloser, options ...TorrentSpecOption) (*TorrentSpec, error) {}

// NewFromMagnet create a torrent from a magnet URL.
func NewFromMagnet(s string, options ...TorrentSpecOption) (*TorrentSpec, error) {}
```

-  Add two new methods to the client that just deal with torrent specs.
```golang
func (cl *Client) AddTorrent(t *TorrentSpec) (new bool, T *Torrent, err error) {}
func (cl *Client) DropTorrent(t *TorrentSpec) error {}
```
- possible add a third method to the client to wait for a specific torrent to complete. I'm still undecided on this one, think it might be better on the torrent struct itself.
```golang
func (cl *Client) AwaitCompletion(c context.Context, t *TorrentSpec) (chan error) {}
```

as an aside, I'm willing to help submit PRs for this work

**Comment by anacrolix** [comment]
I'm not happy with the API either, it has evolved organically, and is in the current state as I haven't decided on a reasonable way forward. I'm leaning toward removing TorrentSpec, and exposing suitable methods on Torrent for anything that can be adjusted dynamically, like trackers. I'm still undecided if it's feasible to make storage and chunk size adjustable while a Torrent is active. A lot of new users of the torrent package will want the shortest path to get up and running, directly adding metainfos, magnet links etc. is sort of ideal for them, so they don't have to deal with any domain complexity initially.

I think a good starting place might be to start pruning down TorrentSpec to just ChunkSize and Storage, and renaming it to TorrentOpts or something.

**Comment by anacrolix** [comment]
I had another read of https://github.com/anacrolix/torrent/issues/218#issuecomment-354567367. Perhaps you could argue a case for why DropTorrent should be on Client, and why TorrentSpec or Torrent should be preferred for options, those are API choices I've struggled with.

**Comment by oliverpool** [comment]
> DropTorrent should be on Client

Since the `Torrent` has a reference to the `Client`, it shouldn't be hard to add a `Drop` method on the torrent, which calls the `DropTorrent` on the client with the appropriate parameters.
This also concerns the other methods.

> I think a good starting place might be to start pruning down TorrentSpec to just ChunkSize and Storage, and renaming it to TorrentOpts or something.

Maybe split what is absolutely necessary (`TorrentBase`) and what is nice-to-have (`TorrentInfo`) to start a torrent.
```go
type TorrentBase struct {
  Storage Storage
  ChunkSize int
  More *TorrentInfo // maybe nil if nothing available
}
type TorrentInfo struct {
  Trackers []string
  ...
}
```

**Comment by oliverpool** [comment]
Instead of
```

**Comment by oliverpool** [comment]
Instead of
```go
func (cl *Client) AwaitCompletion(c context.Context, t *TorrentSpec) (chan error) {}
```

I think I would prefer a blocking
```go
func (cl *Client) DownloadTorrent(c context.Context, t *TorrentSpec) error {}
```
The function should return on torrent completion (or on irrecoverable error or on context cancellation).
It's more designed for a one-shot download (than a long-lived client), but should satisfy many needs.

**Comment by james-lawrence** [comment]
> Since the Torrent has a reference to the Client

I actually think this is a mistake, and one of the goals I'd like to see is decoupling the two.

In my mind the torrent should contain the metadata around accessing and storing the data related to the torrent.

And the client is responsible for coordinating peers and downloading parts of the torrent. Going this path lets us remove the global lock for most calls and instead use a torrent specific locking mechanism which will help the library scale to a larger number of torrent by reducing lock contention.

Basically having Start/Stop/Download on the client lets us deal with just the network behaviour around the torrent. Then a separate Delete method on the torrent for cleaning up the metadata.

I'm pretty indifferent on the function signature of the download function but I do believe `DownloadTorrent` as a name is a poor choice as it'll be confused with `AddTorrent`.

**Comment by james-lawrence** [comment]
> I had another read of #218 (comment). Perhaps you could argue a case for why DropTorrent should be on Client, and why TorrentSpec or Torrent should be preferred for options, those are API choices I've struggled with.

So basically I'm pretty strongly interested in getting rid of one or the other as well. But the structure exposed really should just be related to storage and metadata about the torrent. at which point once we get it down to one exposed type we can just name it Torrent and be done with it.

internally the client would have a private representation of the torrent for handling downloading parts from the DHT.

As a result we'd probably need a `CancelTorrent` on the client and a `Drop` on the torrent. `client.CancelTorrent` stops all network activity and interactions with the DHT for the specific torrent, and `torrent.Drop` will remove and data from storage.

**Comment by james-lawrence** [comment]
As an aside, last weekend I took a stab at this work and have a branch that does a good chunk of the decoupling the torrent from the client regarding locks and moving the methods off the client for creating a torrent spec. all the tests passed, but have yet to actually try it out with a real torrent. can push that up to github if you guys want to take a look.

**Comment by oliverpool** [comment]
> internally the client would have a private representation of the torrent for handling downloading parts from the DHT.

If I understand correctly, the client could then have separated parts:
- one part responsible for the pieces to seed/download (it only needs the hashes and some storage)
- one part responsible for read a .torrent file (and passing the hashes of the different pieces to the previous part)
- one part responsible for fetching the hashes, from a .magnet URL
- ...


Or maybe it is more about layers:
- a network transport layer
- a lower layer dealing with the different pieces (identified by hash)
- an intermediate layer dealing with the individual files (looking for their pieces to be downloaded by the previous layer)
- an upper layer dealing with the torrent (waiting for all the files of the torrent to be downloaded)
- a client dealing with multiple torrents and their completion

Do you think such layer structure would work? (Since I don't know much about the way torrent works, what I wrote might just be stupid)

**Comment by anacrolix** [comment]
@oliverpool @james-lawrence consider discussing in gitter, https://gitter.im/anacrolix/torrent

**Comment by james-lawrence** [comment]
@oliverpool I'll be in the glitter channel for the next couple days if you want to chat.

**Comment by anacrolix** [comment]
@james-lawrence @oliverpool I appreciate the efforts to tidy up the API. I'll try to layout my view from the discussion above and elsewhere.

I want to remove the duplication caused by `TorrentSpec`. The reason for its existence is for options/configuration that couldn't apparently be done dynamically, after a `Torrent` is created on a `Client`. In hindsight, probably everything on the `TorrentSpec` can be done at runtime, and should be for those use-cases like typical GUI torrent clients (adding/removing trackers, adjusting bandwidth usage, etc. on the fly). I don't like the option-style (like `AddTorrent(infoHash, opts...AddTorrentOption)`, but it's preferable to breaking backward compatibility when used right I think. It would be satisfactory for configuration that shouldn't be dynamic on a `Torrent`, perhaps storage would fit into this category. I think the way forward here would be just to remove `TorrentSpec`, and figure out what can be done dynamically on a `Torrent` from there.

I don't want duplication on closing between `Client` and `Torrent`. `Drop` or `Close` should exist on one or the other.

I don't want `Download`, or `DownloadTorrent` stuff. `Client`s and torrents in general don't just have to be about downloading.

The definite issue in this thread is `TorrentSpec`. If you're keen @james-lawrence, an issue and PR removing `TorrentSpec`, moving as much of its configuration to be dynamic methods on `Torrent`, and the rest as `AddTorrentOption`s would seem appropriate. Let me know what you think.

**Comment by james-lawrence** [comment]
@anacrolix - first off thanks for the feedback, much appreciated.

so I think the main issues you have w/ #340 can be summed up with:
1. remove the [Download method](https://github.com/anacrolix/torrent/pull/340/files#diff-d0b7f86cea13f90d3cc0cc2338da8d5fR110) function. I'm perfectly okay with this, it was a nice to have for me. but its not necessary, and we can discuss its benefits separately.
2. You want [Torrent](https://github.com/anacrolix/torrent/pull/340/files#diff-81f624f2512d3b5b7ef5ae1f3a3a85d3R51) and [Download interface](https://github.com/anacrolix/torrent/pull/340/files#diff-88dd413390759db7ad32c2b7380043b3R41) merged together as a single type.

Merging Torrent and Download in #340 is the most problematic part for me. In theory I'm not strictly opposed to it, but it makes the cleanup harder over the next couple of iterations, if you look at the Download interface I've marked a ton of changes I think can occur to simplify the API further. unfortunately combining the two types makes it very unclear when/if those simplifications can occur.
It also leaves confusion on when certain methods can be invoked.

Currently there are methods on would be on Download that if invoked before a Torrent is added to a client can panic/error/deadlock. this is actual issue in master.

for example I believe the current Download interface can be simplified a ton from:
```golang
type Download interface {
	Torrent() Torrent
	Name() string                // TODO: remove, should be pulled from Torrent()
	Metainfo() metainfo.MetaInfo // TODO: remove, should be pulled from Torrent()
	BytesCompleted() int64       // TODO: maybe should be pulled from torrent, it has a reference to the storage implementation.
	VerifyData()                 // TODO: maybe should be pulled from torrent, it has a reference to the storage implementation.
	AddClientPeer(cl *Client)    // TODO: remove, this should only be used in tests.
	NewReader() Reader           // TODO: maybe should be pulled from torrent, it has a reference to the storage implementation.
	Seeding() bool               //
	Stats() TorrentStats         //
	Info() *metainfo.Info        // TODO: remove, this should be pulled from Torrent()
	GotInfo() <-chan struct{}    // TODO: remove, torrents should never be returned in they don't have the meta info.
	DownloadAll()                // TODO: rethink this. does it even need to exist or can it be rolled up into Start/Download.
	Files() []*File              // TODO: maybe should be pulled from torrent, it has a reference to the storage implementation.
	AddPeers(pp []Peer)
	SubscribePieceStateChanges() *pubsub.Subscription
	PieceStateRuns() []PieceStateRun
	SetMaxEstablishedConns(max int) (oldMax int)
}
```
to

```golang
type Download interface {
	Torrent() Torrent    // has the metadata methods, hash, display name, name, etc.
       Content() Storage // has the methods equivalent to current Files(), NewReader(), VerifyData() etc.
       Status() TorrentStats // has the Seeding data, current progress, etc.
       Tune(options ...tuning) error // provide the functionality for tuning an ongoing torrent, rate limits, maximum connections, etc.
}
```

**Comment by james-lawrence** [comment]
finally: if your okay with keeping the types separate until we get a little further along and know more, I can have #340 reworked this weekend.

**Comment by anacrolix** [comment]
I'm not okay with `Download`. The `Torrent` type is about more than just downloading.

**Comment by anacrolix** [comment]
I'm closing this issue in favour of https://github.com/anacrolix/torrent/issues/369, https://github.com/anacrolix/torrent/issues/370, and #168 which are much more specific and approachable. Open a new issue for anything identified in this discussion that hasn't been addressed or tracked elsewhere.

**Comment by jaredallard** [comment]
@anacrolix I think it'd make sense to have a separate issue for context "plumbing" into this API as it stands, e.g. for `client.WaitAll(ctx context.Context)`. There's #168, but that isn't really geared towards that as much as it's specific to one. The greater focus should be adding context cancellation support, for a new one. Happy to create that if wanted.

**Comment by anacrolix** [comment]
@jaredallard feel free to open an issue or PR for that one.
